### PR TITLE
GetNextAssertion

### LINKED ISSUE #196: OpenSK passes the conformance tool
This issue will be used to track the different runs of the official conformance tool from FIDO Alliance against commits of OpenSK.

### PR DESCRIPTION
**Pull request by kaczmarczyck**
Fixes #196.

This PR implements the GetNextAssertion. This includes saving the command state in GetAssertion, and discarding it on timeout or intersecting commands. The correct behaviour for timeouts landed in CTAP 2.1, and should be backwards compatible to 2.0.

Added @jmichelp for the CTAP changes and double-checking with the conformance tool.
Added @ia0 for reviewing the storage changes.

- [x] Tests pass

### CODE REVIEW

File: src/ctap/mod.rs [hunk 1]
```diff
@@ -167,18 +189,26 @@ where
...
 
-    pub fn check_disable_reset(&mut self, timestamp: Timestamp<isize>) {
-        if timestamp - Timestamp::<isize>::from_ms(0) > Duration::from_ms(RESET_TIMEOUT_MS) {
-            self.accepts_reset = false;
+    pub fn update_command_permission(&mut self, now: ClockValue) {
+        self.stateful_command_permission = self.stateful_command_permission.check_expiration(now);
+    }
+
+    fn check_command_permission(&mut self, now: ClockValue) -> Result<(), Ctap2StatusCode> {
+        self.stateful_command_permission = self.stateful_command_permission.check_expiration(now);
```
Comment (jmichelp) [inline src/ctap/mod.rs:207]: why not calling `self.update_command_permission(now);` here?

File: src/ctap/mod.rs [hunk 2]
```diff
@@ -307,20 +343,35 @@ where
...
                     }
+                    Command::AuthenticatorGetNextAssertion => self.process_get_next_assertion(now),
                     Command::AuthenticatorGetInfo => self.process_get_info(),
                     Command::AuthenticatorClientPin(params) => self.process_client_pin(params),
-                    Command::AuthenticatorReset => self.process_reset(cid),
+                    Command::AuthenticatorReset => self.process_reset(cid, now),
                     #[cfg(feature = "with_ctap2_1")]
                     Command::AuthenticatorSelection => self.process_selection(cid),
-                    // TODO(kaczmarczyck) implement GetNextAssertion and FIDO 2.1 commands
-                    _ => self.process_unknown_command(),
```
Comment (jmichelp) [inline src/ctap/mod.rs:323]: Why removing this?
If the command is unsupported we must reply.

Comment (kaczmarczyck) [reply src/ctap/mod.rs:323]: We do return an error_code in `Command::deserialize`. We needed `process_unknown_command` before when we correctly parsed the inputs of some commands, but didn't actually implement them. From now on, I don't expect that we will have a command that we parse, but don't handle here.

File: src/ctap/mod.rs (hunk 1 above)
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: Done

**Review by ia0 — DISMISSED** [review]
LGTM but I have a few questions regarding the semantics of `creation_order` that might require change:
- Should different credentials have different creation order? Currently all existing credentials have a creation order of 0.
- Should the order actually be creation order time? Or would any stable order for the same set of credentials be enough? In store v2, credentials have unique stable identifiers.
- Should different credentials across time have different creation order? Currently adding a credential, removing it, then adding a new one, would share the same creation order.
- (depends on previous questions but) Was a monotonic creation order counter considered?

**Comment by kaczmarczyck** [comment]
To answer your questions:
- The `creation_order` is used for sorting, so non-uniqueness means we can't correctly sort. For existing credentials, we don't know their order of creation, we only know that they are earlier.
- The exact clock value is not important, but we need to know the order in which they are created. So a unique identifier isn't good enough.
- Removed credentials can never be discovered in `GetAssertion`, and we can still sort all found credentials.
- I did consider monotonic counter. It shouldn't be a big difference: We store one more integer, and it's a bit easier to query, since we don't have to compute a max over all credentials. On the other hand, we technically run out of integers a little bit faster, since there is a possibility that removing credentials through credential management (2.1 command) frees up some numbers, and frequent creation and deletion doesn't hurt that much.

We could even filter by `rp_id` for these integers, since we will only have to sort credentials of the same RP ID. Makes caching harder, but keeps numbers lower. Probably no big difference in practise though?

File: src/ctap/mod.rs (hunk 2 above)
Comment (jmichelp) [reply src/ctap/mod.rs:323]: I see. Then we're good.

**Review by jmichelp — APPROVED** [review]

File: src/ctap/mod.rs [hunk 3]
```diff
@@ -619,92 +756,74 @@ where
...
-
+        // This check comes before CTAP2_ERR_NO_CREDENTIALS in CTAP 2.0.
+        // For CTAP 2.1, it was moved to a later protocol step.
         if options.up {
             (self.check_user_presence)(cid)?;
         }
 
-        self.increment_global_signature_counter()?;
+        let (credential, remaining_credentials) = credentials
+            .split_last()
```
Comment (ia0) [inline src/ctap/mod.rs:785]: This seems costly. You convert back the remaining credentials to a `Vec` later. Why not use `pop` here?

**Comment by ia0** [comment]
Ok I think I see. Let me try to recap and if it's good then it's good on my side (modulo the small performance issue).

- We just need to sort credentials with the same RP ID by creation order (updating a credential is like a creation, it just happens that it also deletes the old credential).
- We actually sort credentials by creation order regardless of RP ID.
- We just need to sort credentials at a given point in time (because we can't break a chain of GetNext without "resetting" it, so in particular we can't change the credential states within a loop of GetNext).
- The creation order is just an time-dependent injection of the actual creation time of credentials into `u64` preserving order.

**Comment by kaczmarczyck** [comment]
How to update a credential is a good question, and the specification does not explicitly mention how this case is handled with respect to ordering. It might be more in the spirit of the specification to bump, and that's what's happening.

So: Yes to all points in your summary. Do you have an opinion about this one?

> We actually sort credentials by creation order regardless of RP ID.

We currently just `+ 1` the current maximum, I'll make it a more explicit `wrapping_add`. That would mean that ordering is still correct for all credentials created between to overflows. And we can still cache the current maximum if we ever feel like it.

**Comment by ia0** [comment]
I don't have opinion about RP ID extra filtering. The 2 sides I see are:
- Our current solution (order all credentials) seems more robust to future changes (in case we can list all credentials in creation order).
- Extra filtering seems more robust regarding wraparounds (not sure it's an actual concern).

Good point about making the wrapping explicit. This could have failed in tests or fuzzing, while working correctly in prod.

Also I forgot to add to what I understand that we return in reverse creation order (most recent first). It doesn't seem to impact the choices we have though.

**Comment by kaczmarczyck** [comment]
Yes to most recent first. Background is probably that the browser will show you a list of user names and you choose which one you'd like. And you find the most interesting ones easiest.

I also like the current version best, so we're good.

File: src/ctap/mod.rs (hunk 3 above)
Comment (kaczmarczyck) [reply src/ctap/mod.rs]: Leftover from an old revision. Now using `pop`, and also moving the credential later to avoid a few `clone`s later.

**Review by jmichelp — APPROVED** [review]

**Pull request merged**